### PR TITLE
Naive implementation of client negotiable decoder

### PR DESCRIPTION
**Pull request by 13h3r**
Related issue #16813. This PR is for feedback.
## Implementation

I am not very familiar with HTTP details and I want to confirm that idea of universal negotiable decoder is viable and correct
## API

I am not sure about creating `ResponseBuilding` as analog of `RequestBuilding` (suggested [here](https://github.com/akka/akka/issues/16813#issuecomment-150221382)). My idea is to provide universal building block in terms of functions and flows. Should we push this decoder into _Request Level API_ flow?
## Tests

I really have no idea what to test in this implementation. All tests looks like an implementation duplication.  Only viable idea is to build whole pipeline of response parsing and decoding.

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http/src/main/scala/akka/http/scaladsl/coding/NegotiableDecoder.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  private def findCodec(encoding: HttpEncoding) = encoding match {
+    case HttpEncodings.gzip | HttpEncodings.`x-zip` => Gzip
+    case HttpEncodings.deflate => Deflate
+    case _ => NoCoding
+  }
+
+  def decode(resp: HttpResponse) = findCodec(resp.encoding).decode(resp)
+
+  def flow() = Flow[HttpResponse].map(decode)
+}
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/NegotiableDecoder.scala]: In general I guess it'll be something around that style, yeah.
We should also allow picking a specific one in the same API.
Maybe we could make it more fluent, as in `via.(Coding.decode: Flow)`?

Comment (drewhk) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/NegotiableDecoder.scala]: The flow does not need to be a def, it is supposedly reusable.

**Comment by 13h3r** [comment]
ping

File: akka-http/src/main/scala/akka/http/scaladsl/coding/NegotiableDecoder.scala [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package akka.http.scaladsl.coding
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/NegotiableDecoder.scala]: needs license header (see other files)

**Comment by ktoso** [comment]
Hi @13h3r, sorry I didn't review this last week (much travel)!
As for testing it, have you seen [CodingDirectivesExamplesSpec](https://github.com/akka/akka/blob/release-2.3-dev/akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/directives/CodingDirectivesExamplesSpec.scala) it shows how such code can be tested (see `compress("Hello", Gzip)`). You'll need to write a tests that requests a compressed data (use a server that responds with compressed "Hello"), and on the client side you'd apply the decompression you're writing, most likely like so: `httpResponses.via(NegotiableDecoder.flow).runWith(Sink.head)` and then assert on the entity).

Hope this helps! Feel free to ask for help if stuck, I'm available again so should be able to help out :)

**Comment by ktoso** [comment]
Hi @13h3r, any interest to take this further? :)

**Comment by 13h3r** [comment]
Hey, @ktoso. I have an interest, but right now don't have a lot of time :( I will continue this later. Thanks for ideas and help.

**Pull request closed without merging**